horace_core/spaghetti_plot.m

```diff
@@ -109,426 +113,486 @@
+[args, opt, present] = parse_arguments(varargin, arglist, flags);
 
-if numel(args)~=2
-    if length(args{1})>1 && (isa((args{1}(1)),'d2d') || isa((args{1}(1)),'IX_dataset_2d'))
-        plot_dispersion(args{1},opt);
-        return
-    else
-        error('SPAGHETTI_PLOT:invalid_arguments',...
-            'Invalid number of arguments')
-    end
+if numel(args) == 1 && isa(args{1}(1), 'd2d') || isa(args{1}(1), 'IX_dataset_2d')
+    plot_dispersion(args{1}, opt);
+    return
+elseif isa(args{2}(1), 'd2d')
+    plot_dispersion(args{2}, opt);
+    return
+end
+
+if numel(args) ~= 2
+    error('SPAGHETTI_PLOT:invalid_arguments', ...
+          'Invalid number of arguments')
 end
 
-if size(args{1},2)~=3 || size(args{1},1)<2
-    error('SPAGHETTI_PLOT:invalid_arguments',...
-        'It should be at least 2 rlp arranged in array [Nx3] (N>=2) but size of the rlp array is: [%d,%d];',...
-        size(args{1}))
+rlp = args{1};
+num_rlp = size(rlp, 1);
+nseg = num_rlp - 1;
+sqw_in = args{2};
+
+if size(rlp, 2) ~= 3 || num_rlp < 2
+    error('SPAGHETTI_PLOT:invalid_arguments', ...
+          'It should be at least 2 rlp arranged in array [Nx3] (N >= 2) but size of the rlp array is: [%d, %d];', ...
+          size(rlp))
 end
+
 if present.cuts_plot_size
-    if size(args{1},1)-1 ~= numel(opt.cuts_plot_size)
-        error('SPAGHETTI_PLOT:invalid_arguments',...
-            [' the size of the cut_plot_dist array should be one less than the number of rlp points.'...
-            ' In fact the number of rlp is %d and the number of distances is %d'],...
-            size(args{1},1),numel(opt.cuts_plot_size));
-    else
-        if any(opt.cuts_plot_size<=0)
-            invalid = opt.cuts_plot_size<=0;
-            error('SPAGHETTI_PLOT:invalid_arguments',...
-                'the plot sizes should be positive numbers but some of them are: %g; %g; %g; %g; %g',...
-                opt.cuts_plot_size(invalid));
-        end
+    if nseg ~= numel(opt.cuts_plot_size)
+        error('SPAGHETTI_PLOT:invalid_arguments', ...
+              [' the size of the cut_plot_dist array should be one less than the number of rlp points.'...
+               ' In fact the number of rlp is %d and the number of distances is %d'], ...
+              num_rlp, numel(opt.cuts_plot_size));
     end
-end
 
-try
-    sqwfile = exist(args{2},'file');
-catch
-    sqwfile = 0;
+    invalid = opt.cuts_plot_size <= 0;
+    if any(invalid)
+        error('SPAGHETTI_PLOT:invalid_arguments', ...
+              'the plot sizes should be positive numbers but some of them are: %g; %g; %g; %g; %g', ...
+              opt.cuts_plot_size(invalid));
+    end
 end
 
-if isa(args{2},'sqw')
-    header = struct(args{2}).data;
-elseif sqwfile
-    header = head_sqw(args{2});
-elseif length(args{2})>1 && isa(args{2}(1),'d2d')
-    plot_dispersion(args{2},opt);
-    return;
+if isa(sqw_in, 'sqw')
+
+elseif istext(sqw_in) && is_file(sqw_in)
+    sqw_in = sqw(sqw_in);
```

Comment:
Hmmm... I think in most cases (especially with data from MERLIN or LET) this will not be an issue because the full `sqw` file will be far too big to fit into memory. However, for shorter runs or for data from MAPS (or maybe SEQUOIA) then you might conceivably get data which just fits into memory. This would then read in all that data which will make the whole computation slower when it doesn't need to be. I guess that as `sqw_in` goes out of scope when the function exits that Matlab should then clean up the memory so the user will not run out of memory but it still seems a waste to read everything in when we will call `cut` on it just a few lines below...

Is there a way to force it to use file-backing (e.g. just load the headers and not pixels)?

Reply:
That is what it does, I believe, under the new file-backing scheme or at least that was the intent.

Reply:
Ah so `sqw(filename)` always just loads it file-backed (e.g. just the headers loaded, not pixels)?

Reply:
For small files below `mem_chunk_size*fb_scale_factor` pixels yes. I can forcibly make it always load file-backed even for those small files if you prefer.

Reply:
Ah, ok, so it just loads one "page"... I think it's ok... but just to check - is the numbers here in bytes or MB? (On my system `mem_chunk_size` is `5e6` and `fb_scale_factor` is `10` so it would either load the file if it was 50MB (fine) or 50GB (not so good, will take a long-ish time)...

Reply:
`Pixels` or `bytes`, I can't remember which we ended up going for. I think pixels.

Reply:
So IDAaaS is back up and I just checked - it the chunk size is the number of **pixels**.

For "small" machines the [default](https://github.com/pace-neutrons/Horace/blob/b77e7e5633709a1cf03d13bd7dc692d215491cc4/herbert_core/admin/%40opt_config_manager/OptimalConfigInfo.xml#L90) `mem_chunk_size=5e6`; for ["large" machines](https://github.com/pace-neutrons/Horace/blob/b77e7e5633709a1cf03d13bd7dc692d215491cc4/herbert_core/admin/%40opt_config_manager/OptimalConfigInfo.xml#L118) it is `mem_chunk_size=20e6`, and `fb_scale_factor` is always `10`.

Given that 1 pixel is [9 doubles](https://github.com/pace-neutrons/Horace/blob/master/documentation/add/07-sqw_redesign.md#pixel-block) this means around 72 bytes per pixel (without overheads), so for "large" configurations the largest amount of memory used would be ~14GB whilst for a "small" configuration it would be ~3.5GB...

I reckon this is ok, there's unlikely to be a full `sqw` (even for MAPS) which is less than 20GB... so I don't think we need to change anything in this PR.

It might be an idea to tweak either the `mem_chunk_size` or `fb_scale_factor` values in future though...